corgidrp/photon_counting.py

```diff
@@ -0,0 +1,353 @@
+
+    The frames in each stack should be SCI image (1024x1024) frames that have 
+    had some of the L2b steps applied:
+
+    - have had their bias subtracted
+    - have had masks made for cosmic rays
+    - have been corrected for nonlinearity (These first 3 steps make the frames L2a.)
+    - have been frame-selected (to weed out bad frames)
+    - have been converted from DN to e-
+    - have been desmeared if desmearing is appropriate.  Under normal
+    circumstances, these frames should not be desmeared.  The only time desmearing
+    would be useful is in the unexpected case that, for example,
+    dark current is so high that it stands far above other noise that is
+    not smeared upon readout, such as clock-induced charge, 
+    fixed-pattern noise, and read noise.  For such low-exposure frames, though,
+    this really should not be a concern.
+
+    This algorithm will photon count each frame in each of the 2 sets (illuminated and dark) individually,
+    then co-add the photon counted frames. The co-added frame is then averaged
+    and corrected for thresholding and coincidence loss, returning the mean
+    expected array in units of photoelectrons.  The threshold is 
+    determined by "T_factor" stored in DetectorParams. The dark mean expected array
+    is then subtracted from the illuminated mean expected array.
+
+    Args:
+        input_dataset (corgidrp.data.Dataset): This is an instance of corgidrp.data.Dataset containing the 
+            photon-counted illuminated frames as well as the 
+            photon-counted dark frames, and all the frames must have the same 
+            exposure time, EM gain, k gain, and read noise.
+        detector_params (corgidrp.data.DetectorParams): A calibration file storing detector calibration values.
+        niter (int, optional): Number of Newton's method iterations (used for photometric 
+            correction). Defaults to 2.
+
+    Returns:
+        pc_dataset (corgidrp.data.Dataset): Contains mean expected photoelectron array (dark-subtracted)
+
+    References
+    ----------
+    [1] https://www.spiedigitallibrary.org/conference-proceedings-of-spie/11443/114435F/Photon-counting-and-precision-photometry-for-the-Roman-Space-Telescope/10.1117/12.2575983.full
+    [2] https://doi.org/10.1117/1.JATIS.9.4.048006
+
+    B Nemati, S Miller - UAH - 13-Dec-2020
+    Kevin Ludwick - UAH - 2023
+
+    """
+    test_dataset, _ = input_dataset.copy().split_dataset(exthdr_keywords=['EXPTIME', 'CMDGAIN', 'KGAIN', 'RN'])
+    if len(test_dataset) > 1:
+        raise PhotonCountException('All frames must have the same exposure time, '
+                                   'commanded EM gain, and k gain.')
+    datasets, vals = test_dataset[0].split_dataset(prihdr_keywords=['VISTYPE'])
+    if len(vals) != 2:
+        raise PhotonCountException('There should only be 2 datasets, and one should have \'VISTYPE\' = \'DARK\'.')
+    dark_dataset = datasets[vals.index('DARK')]
```

Comment:
I thought the only reason we made the synthesized and traditional master darks into calibration products was because they were CTC deliverables, so I didn't make the photon-counted dark into a separate calibration product because I don't think it is.  Is this accurate, @mygouf ?